docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,212 @@
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer and fee payer 
+the transaction, the `priority` could be the cumulative sum of the fee(s).
+
+Only `sender` is required, while `priority` can be omitted which would result in
+using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:*Tx>`, i.e a mapping from `sender` to a single `*Tx`
+object, where each `*Tx` is the next valid and processable transaction from the
+given `sender`.
+
+On top of this mapping, we index all transactions by priority using a thread-safe
+priority queue, i.e. a [max heap](https://en.wikipedia.org/wiki/Min-max_heap).
+When a proposer is ready to select transactions for the next block proposal,
+transactions are selected from this priority index by highest priority order.
+When a transaction is selected and reaped, it is removed from this index and
+from the `<sender:*Tx>` mapping.
+
+We define `Tx` as the following data structure:
+
+```go
+type Tx struct {
+  // Tx represents the raw binary transaction data.
+  Tx []byte
+
+  // Priority defines the transaction's priority as specified by the application
+  // in the ResponseCheckTx response.
+  Priority int64
+
+  // Sender defines the transaction's sender as specified by the application in
+  // the ResponseCheckTx response.
+  Sender string
+
+  // Index defines the current index in the priority queue index. Note, if
+  // multiple Tx indexes are needed, this field will be removed and each Tx
+  // index will have its own wrapped Tx type.
+  Index int
+}
+```
+
+### Eviction
+
+Upon successfully executing `CheckTx` for a new `Tx` and the mempool is currently
+full, we must check if there exists a `Tx` of lower priority that can be evicted
+to make room for the new `Tx` with higher priority and with sufficient size
+capacity left.
+
+If such a `Tx` exists, we find it by obtaining a read lock and sorting the
+priority queue index. Once sorted, we find the first `Tx` with lower priority and
+size such that the new `Tx` would fit within the mempool's size limit. We then
+remove this `Tx` from the priority queue index as well as the `<sender:*Tx>`
+mapping.
+
+This will require additional `O(n)` space and `O(n*log(n))` runtime complexity.
+
+### Gossiping
+
+When gossiping new transactions to peers, we first obtain a read lock on the
+priority queue index. If we have not gossiped the top priority `Tx` for the
+given peer, we return this `Tx` as the next transaction to gossip. Otherwise,
+we sort the priority queue index and traverse in descending priority order until
+we find the first non-sent `Tx` to the peer.
```

Comment:
Do we keep track of what tx's we have already gossiped to each peer?

Reply:
Yes, that is the cache that I mention below.

Reply:
Maybe I'm wrong, but this strikes me as being quite costly. If we are connected to 50 peers, we're going to have 50 threads all holding various read locks on the priority queue plus having a cache for each peer. I'm trying to think of a better construction but haven't found one yet. Batching tx's would certainly help.

Reply:
Also a characteristic with priority sending is that most peers will all be sending one another the same txs. I guess that can't be avoided

Reply:
One thought I had was to distinguish whether the tx was submitted to you directly or whether you received it via gossiping. I believe users generally submit to a single node therefore the node should first try to sends txs it receives through the rpc to all the peers it's connected to and then if there aren't any more to send, go back to regossiping the txs that were gossiped to you

Reply:
> Maybe I'm wrong, but this strikes me as being quite costly. If we are connected to 50 peers, we're going to have 50 threads all holding various read locks on the priority queue plus having a cache for each peer. I'm trying to think of a better construction but haven't found one yet. Batching tx's would certainly help.

Goroutines are cheap. There will only be a single lock, but yes, there will be various bits of contention between reads and writes amongst the indexes and gossiping. Happy to hear alternatives :)

> Also a characteristic with priority sending is that most peers will all be sending one another the same txs. I guess that can't be avoided

This already happens AFAIK.

Reply:
I think it's fine to leave it as is and focus on it being functional and then at a later point we can review methods that might be more performant. I still like my idea about having a distinction between txs you receive from users and tx you receive from other nodes

Reply:
Not quite understand how the tx Gossiping work in Tendermint Core. Is it better for every new tx coming into the mempool, if the transaction is valid (ready for executing) and then gossip (or say broadcast) to the connecting peers. Eventually, the whole network should have the same tx replica. Maybe there are few nodes that don't have the tx. It is still okay for proposing a block, right? Do we need to do another gossip for the missing high-priority txs? So, in this way we don't need to know what txs the connecting peers have.

Reply:
> One thought I had was to distinguish whether the tx was submitted to you directly or whether you received it via gossiping.

I believe we already do this through the TxInfo - an empty TxInfo indicates the tx came from the RPC, otherwise it will have the id of the peer we received from.